src/services/task-scheduler-service.ts

```diff
@@ -47,22 +54,37 @@ export class TaskSchedulerService {
     )
 
     this._scheduledTask$ = taskWithRetryOnError$.pipe(
-      expand(() => {
+      expand((success: boolean) => {
         if (this._controller.signal.aborted) {
           return EMPTY
         }
 
+        if (cbAfterFailure && !success) {
+          logger.imp(`Last run of the task was not successful. Stopping the task scheduler`)
+          return EMPTY
```

Comment:
Will this cause an exit after the very first time a batch isn't found? Or will it run through the extra attempts and then exit if each attempt fails?

Reply:
It actually doesn't matter a lot if it exits after the first attempt. As long as we're making an attempt to check the queue before shutting down.

Reply:
So there are two "loops" in this situation.

1. The task is anchoring a queued batch. Anchoring a queued batch has a loop when retrieving a task. 
2. The task scheduler service which runs the task after a delay of 1000s 

Originally I added #1 because for our docker environment we wanted to run the workers in "anchor" mode and have the worker wait for a message. So I made waiting for a message run in a loop. 

I then added #2 because we wanted to do greedy anchoring. So we run #1 in a loop with a dealy

This didn't really work with our infra because #1 caused use to wait indefinitely for a message which meant long living workers. So i think i'm just going to remove #1 because its confusing, and rely on the task scheduler

Reply:
Ok, I see, ty for explaining.

So, IIUC, now we're relying on the existing task/loop code to attempt to retrieve queue messages. With a 10 second wait time, the call to SQS will wait up to that long for a batch. If it finds one, it will process the batch, then look for more batches. If not, that will be considered a "failure" of which a max of 3 attempts will be permitted before a final failure is detected and the callback is invoked, exiting the process.

Reply:
I definitely like the simplicity of the new approach!!

Reply:
Almost. 

We are retrying on the existing task/loop code. This runs the task. The task is wait for 10s for a batch then:
1. if there is a batch anchor it. The task will then complete returning true. We then wait for a delay of schedulerIntervalMS before starting again
2. if there isn't a batch complete and return false indicating a no op. At that point the scheduler will invoke the cb exiting the process.

Its hard to get the the right terminology. The retry logic is only invoked when errors are thrown, not if the task returns false. If you have a better name please let me know

Reply:
> And if we want to do long-lived workers in the future, we could make the retry count configurable and set it to infinity?

If we wanted long lived workers we would set schedulerStopAfterFail to false

Reply:
Yeah, I like the differentiation between no-op and failure, otherwise it gets confusing.

Reply:
> if there isn't a batch complete and return false indicating a no op. At that point the scheduler will invoke the cb exiting the process.

How does this retry factor in there?
```
      retry({
        delay: intervalMS,
        count: 3,
```

Reply:
> > if there isn't a batch complete and return false indicating a no op. At that point the scheduler will invoke the cb exiting the process.
> 
> How does this retry factor in there?
> 
> ```
>       retry({
>         delay: intervalMS,
>         count: 3,
> ```

The retry is only if the task throws errors. Ex. The scheduler starts, the task throws an error, we retry. If it throws an error again we retry it etc. If we have retried it 3 times, the scheduler will throw causing the entire thing to stop